contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+against all the nodes.  For each node, keep an in-memory map of all its PVs
+keyed by StorageClass.  When finding matching PVs for a particular node, try to
+match against the PVs in the node’s PV map instead of the cluster-wide PV list.
+
+For the alpha phase, the optimizations are not required.  However, they should
+be required for beta and GA.
+
+#### Packaging
+The new bind logic that is invoked by the scheduler can be packaged in a few
+ways:
+* As a library to be directly called in the default scheduler
+* As a scheduler extender
+
+We propose taking the library approach, as this method is simplest to release
+and deploy.  Some downsides are:
+* The binding logic will be executed using two different caches, one in the
+scheduler process, and one in the PV controller process.  There is the potential
+for more race conditions due to the caches being out of sync.
+* Refactoring the binding logic into a common library is more challenging
+because the scheduler’s cache and PV controller’s cache have different interfaces
+and private methods.
+
+##### Extender cons
+However, the cons of the extender approach outweighs the cons of the library
+approach.
+
+With an extender approach, the PV controller could implement the scheduler
+extender HTTP endpoint, and the advantage is the binding logic triggered by the
+scheduler can share the same caches and state as the PV controller.
+
+However, deployment of this scheduler extender in a master HA configuration is
+extremely complex.  The scheduler has to be configured with the hostname or IP of
+the PV controller.  In a HA setup, the active scheduler and active PV controller
+could run on the same, or different node, and the node can change at any time.
+Exporting a network endpoint in the controller manager process is unprecedented
+and there would be many additional features required, such as adding a mechanism
+to get a stable network name, adding authorization and access control, and
+dealing with DDOS attacks and other potential security issues.  Adding to those
+challenges is the fact that there are countless ways for users to deploy
+Kubernetes.
+
+With all this complexity, the library approach is the most feasible in a single
+release time frame, and aligns better with the current Kubernetes architecture.
+
+#### Other Alternatives Considered
+
+##### One scheduler function
+An alternative design considered was to do the predicate, priority and bind
+functions all in one function at the end right before Pod binding, in order to
+reduce the number of passes we have to make over all the PVs.  However, this
+design does not work well with pod preemption.  Pod preemption needs to be able
+to evaluate if evicting a lower priority Pod will make a higher priority Pod
+schedulable, and it does this by re-evaluating predicates without the lower
+priority Pod.
+
+If we had put the MatchUnboundPVCs predicate at the end, then pod preemption
```

Comment:
This can possibly be addressed by registering extra functions that preemption logic should call to filter nodes further. This might be an easier change compared to the current proposal.

Reply:
Yes I think that would eliminate the need for some of the caching from the current design.  But it will still require a second pass afterwards.

Reply:
That's right. The second pass will still be needed.